### PR TITLE
unrolled transform!

### PR DESCRIPTION
**Pull request by jw3126**
Towards better performance. Roughly 3-4 times faster then baseline, still 2-3 times slower then nettle.

### CODE REVIEW

**Comment by oxinabox** [comment]
How do the `if` statements compare to using a series of loops?
I think the series of `for` loops should be faster

**Comment by jw3126** [comment]
These are compile time ifs. No runtime branches left.

File: src/core.jl [hunk 1]
```diff
@@ -50,6 +51,56 @@ function transform!(context::MD5_CTX)
     @inbounds context.state .+= [a,b,c,d]
 end
 
+@generated function transform_unrolled!(context::MD5_CTX)
+    ret = quote
+        pbuf = buffer_pointer(context)
+    end
+    ex  = quote
+        A = context.state[1]
```
Comment (oxinabox) [inline src/core.jl]: add `@inbounds` to all of these?

**Comment by oxinabox** [comment]
> These are compile time ifs. No runtime branches left.

Ah yes, I just realized that.

File: src/core.jl (hunk 1 above)
Comment (jw3126) [reply src/core.jl]: The whole body is `@inbounds`.

Comment (oxinabox) [reply src/core.jl]: Or does it propagate correctly from the last line?
I'm never sure about `@inbounds` propergation

Comment (jw3126) [reply src/core.jl]: It just does not propagate into function calls. But I think you are still right it is better style to be explict, if something is refactored into a function.

**Comment by oxinabox** [comment]
You should add yourself to the LISCENSE.md

**Comment by oxinabox** [comment]
nettle is hand-optimized assembly.
I'm not sure how bad one can feel about losing to it.
I'm pretty sure it actually manages to not use RAM for anything but the message.
Keeping the A,B,C,D,F all in the CPU registers.

**Comment by jw3126** [comment]
I read [this](https://web.njit.edu/~hz278/md5/report.php) their baseline is similar to ours and the significant optimizations were unrolling and telling C to put variables into registers. (They do use some assembly, but it does not give much performance.) Not sure if there is a way to control register usage from Julia.

**Comment by oxinabox** [comment]
thanks,  this is great
There might be a way to to register forcing via `Core.Intrinsics.llvmcall` but I'm not sure.
It can come in another PR if it matters.

MD5 checking isn't a massive bottleneck ever anyway.
Its basically only for legacy file checksums, right?

**Comment by jw3126** [comment]
Yeah performance is not critical here. My motivation is just curiosity.

**Pull request merged**